File: CCEaseRefresh/CCEaseRefresh/CCEaseRefresh.m
//
//  CCEaseRefresh.m
//  CCEaseRefresh
//
//  Created by v－ling on 15/9/18.
//  Copyright (c) 2015年 LiuZeChen. All rights reserved.
//

#import "CCEaseRefresh.h"
#import <QuartzCore/QuartzCore.h>

#define kRefreshViewHeight 64
#define kTimeViewHeight    15
#define kTimeBottomEdage   8
#define kSubviewEdage      7
#define kBallImage         [UIImage imageNamed:@"refresh_sphere"]
#define kCircelImage       [UIImage imageNamed:@"refresh_circle"]
#define kContentOffset     @"contentOffset"
#define kAttributeDict     @{NSFontAttributeName: [UIFont systemFontOfSize:12.0]}

@interface CCEaseRefresh () {
    CGFloat _ballWidth;     ///< 球的宽
    CGFloat _ballHeight;    ///< 球的高
    CGFloat _circleWidth;   ///< 圆的宽
    CGFloat _circleHeight;  ///< 圆的高
    CGFloat _defaultBallY;  ///< 球的初始Y
    CGFloat _textX;         ///< 文本的X
    CGFloat _textY;         ///< 文本的Y
    CGFloat _currentOffsetY;///< scrollView的当前offset的Y
}

@property (nonatomic, strong) UIScrollView  *scrollView;
@property (nonatomic, strong) CALayer       *ballLayer;
@property (nonatomic, strong) CALayer       *circleLayer;
@property (nonatomic, copy)   NSString      *lastUpdateTimeString;
@property (nonatomic, assign) UIEdgeInsets  originalContentInset;

@end

@implementation CCEaseRefresh

#pragma mark - 初始、释放
- (instancetype)initInScrollView:(UIScrollView *)scrollView {
    self = [super initWithFrame:CGRectMake(0, scrollView.contentInset.top, scrollView.frame.size.width, kRefreshViewHeight)];
    if (self) {
        NSAssert(scrollView, @"CCEaseRefresh's scrollView can't is nil");

        self.scrollView = scrollView;
        self.autoresizingMask = UIViewAutoresizingFlexibleWidth;
        self.backgroundColor = [UIColor clearColor];

        [self.scrollView addObserver:self forKeyPath:kContentOffset options:NSKeyValueObservingOptionNew context:nil];
        [self.scrollView.superview insertSubview:self belowSubview:scrollView];

        [self loadComponent];
    }
    return self;
}

- (void)dealloc {
    printf("CCEaseRefresh(OC) deinit...\n");
    [self.scrollView removeObserver:self forKeyPath:kContentOffset];
    self.scrollView = nil;
}

- (void)willMoveToSuperview:(UIView *)newSuperview {
    [super willMoveToSuperview:newSuperview];
    if (!newSuperview) {
        [self.scrollView removeObserver:self forKeyPath:kContentOffset];
        self.scrollView = nil;
    }
}

- (void)drawRect:(CGRect)rect {
    [self calcTextXY];
    [_lastUpdateTimeString drawAtPoint:CGPointMake(_textX, _textY) withAttributes:kAttributeDict];
}

- (void)calcTextXY {
    if (_textY == 0 || _textX == 0) {
        CGSize textSize = [_lastUpdateTimeString boundingRectWithSize:CGSizeMake(MAXFLOAT, MAXFLOAT) options:NSStringDrawingUsesFontLeading | NSStringDrawingUsesLineFragmentOrigin attributes:kAttributeDict context:nil].size;
        _textY = self.frame.size.height - textSize.height - kSubviewEdage;
        _textX = (self.frame.size.width - textSize.width) / 2;
    }
}

#pragma mark - 加载零件
- (void)loadComponent {
    [self updateTime];
    [self calcTextXY];
    
    // 球和圆View
    _ballWidth = kBallImage.size.width   - kSubviewEdage / 2;
    _ballHeight = kBallImage.size.height - kSubviewEdage / 2;
    _ballLayer = [CALayer layer];
    _ballLayer.frame = CGRectMake((self.frame.size.width - _ballWidth) / 2, -_ballHeight, _ballWidth, _ballHeight);
    _ballLayer.position = CGPointMake(self.center.x, _ballLayer.position.y);
    _ballLayer.contents = (id)kBallImage.CGImage;
    _ballLayer.masksToBounds = YES;
    _circleWidth = kCircelImage.size.width;
    _circleHeight = kCircelImage.size.height;
    _circleLayer = [CALayer layer];
    _circleLayer.frame = CGRectMake((self.frame.size.width - _circleWidth) / 2, 0, _circleWidth, _circleHeight);
    _circleLayer.contents = (id)kCircelImage.CGImage;
    _circleLayer.hidden = YES;
    _defaultBallY = _ballLayer.frame.origin.y;

    // 添加到self、设置透明度
    [self setAlpha:0.f];
    [self.layer addSublayer:_ballLayer];
    [self.layer insertSublayer:_circleLayer below:_ballLayer];
}

#pragma mark - KVO
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    if ([keyPath isEqualToString:kContentOffset]) {
        // offset
        _currentOffsetY = [[change valueForKey:NSKeyValueChangeNewKey] CGPointValue].y;
        // 判断是否可以进入刷新状态
        CGFloat newOffsetThreshold = -_originalContentInset.top;

        if (_refreshState == CCEaseRefreshStateLoading || _currentOffsetY > newOffsetThreshold) {return;}

        if (!self.scrollView.isDragging && _refreshState == CCEaseRefreshStateTrigger) {
            self.refreshState = CCEaseRefreshStateLoading;
        }else if (_currentOffsetY <= newOffsetThreshold * 2 && self.scrollView.isDragging) {
            self.refreshState = CCEaseRefreshStateTrigger;
        }else if (_currentOffsetY < newOffsetThreshold && _currentOffsetY > newOffsetThreshold * 2  && _refreshState != CCEaseRefreshStateLoading) {
            self.refreshState = CCEaseRefreshStateVisible;
        }else if (_currentOffsetY >= newOffsetThreshold && _refreshState != CCEaseRefreshStateDefault && _refreshState != CCEaseRefreshStateLoading) {
            // refreshState != CCRefreshDefault:进入默认状态后,在没有做上拉或下拉刷新时,不给refreshState的setter方法,从而启到优化程序的作用.想让谁少走几次owner != owner.
            self.refreshState = CCEaseRefreshStateDefault;
        }
        return;
    }
}

#pragma mark - set方法群
- (void)setScrollView:(UIScrollView *)scrollView {
    _scrollView = scrollView;
    _originalContentInset = _scrollView.contentInset;
    _scrollView.backgroundColor = [UIColor clearColor];
}

- (void)setRefreshState:(CCEaseRefreshState)refreshState {
    _refreshState = refreshState;
    switch (_refreshState) {
        case CCEaseRefreshStateDefault:
        {
            [self updateTime];
            [self updateContentInset:_originalContentInset];
        }
            break;
        case CCEaseRefreshStateVisible:
        {
            [self updateTime];
            [self updateBallLayerPosition];
        }
            break;
        case CCEaseRefreshStateTrigger:
        {
            [self updateTime];
        }
            break;
        case CCEaseRefreshStateLoading:
        {
            UIEdgeInsets ei = _originalContentInset;
            ei.top = ei.top + kRefreshViewHeight;
            [self updateContentInset:ei];
            [self updateTime];
            [self updateBallLayerPosition];
            [self updatecircleLayerPosition];
            [self sendActionsForControlEvents:UIControlEventValueChanged];
        }
            break;
        default:
            break;
    }
}

#pragma mark - 更新(Layer的Frame、刷新的时间)
- (void)updateBallLayerPosition {
    // fabs(_currentOffsetY) - _originalContentInset.top从0.f开始升序
    CGFloat ballLayerMinY = fabs(_currentOffsetY) - _originalContentInset.top + _defaultBallY;
    CGFloat ballLayerMaxY = _textY - kSubviewEdage - _ballHeight / 2;
    if (ballLayerMinY >= ballLayerMaxY) {
        ballLayerMinY  = ballLayerMaxY;
    }

    // 去掉隐式动画
    // 1.
    [CATransaction begin];
    [CATransaction setValue:(id)kCFBooleanTrue forKey:kCATransactionDisableActions];
    _ballLayer.position = CGPointMake(_ballLayer.position.x, ballLayerMinY);
    [CATransaction commit];
    // 2.
    _ballLayer.actions = @{@"position":[NSNull null]};

    // 更新self的透明度
    self.alpha = (fabs(_currentOffsetY) - _originalContentInset.top) / kRefreshViewHeight;
    self.layer.hidden = NO;
}

- (void)updatecircleLayerPosition {
    _circleLayer.position = CGPointMake(_circleLayer.position.x, _ballLayer.position.y - kSubviewEdage);
    _circleLayer.hidden = NO;
    // 开始动画
    [self ballAnimation];
}

// 更新contentInset
- (void)updateContentInset:(UIEdgeInsets)ei {
    [UIView animateWithDuration:.25 animations:^{
        _scrollView.contentInset = ei;
    }];
}

- (void)updateTime {
    switch (_refreshState) {
        case CCEaseRefreshStateDefault:
        case CCEaseRefreshStateVisible:
            _lastUpdateTimeString = CCEaseDefaultTitle;
            break;
        case CCEaseRefreshStateTrigger:
            _lastUpdateTimeString = CCEaseTriggertTitle;
            break;
        case CCEaseRefreshStateLoading:
            _lastUpdateTimeString = CCEaseLoadingTitle;
            break;
        default:
            break;
    }
    [self setNeedsDisplay];
}

#pragma mark - 动画(CoreAnimation)
- (void)ballAnimation {
    CABasicAnimation *enlarge = [CABasicAnimation animationWithKeyPath:@"transform.scale"];
    enlarge.duration = .5;
    enlarge.repeatCount = 1;
    enlarge.removedOnCompletion = NO;
    enlarge.fromValue = [NSNumber numberWithFloat:.5];
    enlarge.toValue = [NSNumber numberWithFloat:1];
    CABasicAnimation *decrease = [CABasicAnimation animationWithKeyPath:@"transform.scale"];
    decrease.duration = .5;
    decrease.repeatCount = 1;
    decrease.removedOnCompletion = NO;
    decrease.beginTime = .5;
    decrease.fromValue = [NSNumber numberWithFloat:1];
    decrease.toValue = [NSNumber numberWithFloat:.5];
    CABasicAnimation *position = [CABasicAnimation animationWithKeyPath:@"position"];
    position.duration = 1;
    position.repeatCount = MAXFLOAT;
    position.fromValue = [NSValue valueWithCGPoint:_circleLayer.position];
    position.toValue = [NSValue valueWithCGPoint:CGPointMake(_circleLayer.position.x, _circleLayer.position.y + kTimeViewHeight)];
    CAAnimationGroup *group = [CAAnimationGroup animation];
    group.duration = 1;
    group.repeatCount = MAXFLOAT;
    group.removedOnCompletion = NO;
    group.autoreverses = YES;
    group.animations = [NSArray arrayWithObjects:enlarge, decrease, position, nil];
    [_circleLayer addAnimation:group forKey:@"CCEaseRefresh"];
}

#pragma mark - 开始刷新、结束刷新
- (void)beginRefreshing {
    if (_refreshState != CCEaseRefreshStateLoading) {
        self.refreshState = CCEaseRefreshStateLoading;
    }
}

- (void)endRefreshing {
    if (_refreshState != CCEaseRefreshStateDefault) {
        self.refreshState = CCEaseRefreshStateDefault;
        self.alpha = 0.f;
        self.layer.hidden = YES;

        [_circleLayer removeAnimationForKey:@"CCEaseRefresh"];
        [_circleLayer setHidden:YES];
        _ballLayer.position = CGPointMake(_ballLayer.position.x, _defaultBallY);
    }
}

@end

File: CCEaseRefresh/CCEaseRefresh/CCEaseRefresh.swift
//
//  CCEaseRefresh.swift
//  EaseRefresh
//
//  Created by v－ling on 15/9/21.
//  Copyright (c) 2015年 LiuZeChen. All rights reserved.
//

import UIKit

class CCEaseRefresh: UIControl {

    enum CCEaseRefreshState: Int {
        case Normal = 1
        case Visible
        case Targger
        case Loading
    }

    var textX: CGFloat = 0            ///< 文本的X
    var textY: CGFloat = 0            ///< 文本的Y
    var ballWidth: CGFloat = 0        ///< 球的宽度
    var ballHeight: CGFloat = 0       ///< 球的高度
    var circleWidth: CGFloat = 0      ///< 圆的宽度
    var cricleHeight: CGFloat = 0     ///< 圆的高度
    var defaultBallY: CGFloat = 0     ///< 球默认的Y
    var currentOffsetY: CGFloat = 0   ///< 当前的offsetY
    var lastUpdateTimeString: NSString!

    var ballLayer: CALayer!
    var circleLayer: CALayer!
    var textLabel: NSString!
    var targetView: UIScrollView!
    var originalContentInset: UIEdgeInsets!

    let kContentOffset: String = "contentOffset"
    let kRefreshHeight: CGFloat = 64.0
    let kTimeViewHeight: CGFloat = 15
    let kSubviewEdage: CGFloat = 7
    let kBallImage = UIImage(named: "refresh_sphere")!
    let kCircleImage = UIImage(named: "refresh_circle")!
    let kTextAttribute: Dictionary = [NSFontAttributeName: UIFont.systemFontOfSize(12.0)]

    let CCEaseDefaultTitle: String  = "下拉刷新"
    let CCEaseTriggertTitle: String = "松开刷新"
    let CCEaseLoadingTitle: String  = "正在刷新"

    var _refreshState = CCEaseRefreshState.Normal
    var refreshState: CCEaseRefreshState {
        get {
            return _refreshState
        }set {
            _refreshState = newValue
            switch newValue {
            case .Normal:
                self.updateTime()
                self.updateContentOffset(originalContentInset)
            case .Visible:
                self.updateTime()
                self.updateBallLayerPosition()
            case .Targger:
                self.updateTime()
            case .Loading:
                var ei: UIEdgeInsets = originalContentInset
                ei.top = ei.top + kRefreshHeight
                self.updateContentOffset(ei)
                self.updateTime()
                self.updateBallLayerPosition()
                self.updateCircleLayerPosition()
                self.sendActionsForControlEvents(.ValueChanged)
            }
        }
    }

    func updateContentOffset(ei: UIEdgeInsets) {
        UIView.animateWithDuration(0.25) { () -> Void in
            self.targetView.contentInset = ei
        }
    }

    // 更新(Layer的Frame、刷新的时间)
    func updateBallLayerPosition() {
        var ballLayerMinY: CGFloat = fabs(currentOffsetY) - originalContentInset.top + defaultBallY
        let ballLayerMaxY: CGFloat = textY - kSubviewEdage - ballHeight / 2
        if ballLayerMinY >= ballLayerMaxY {
           ballLayerMinY =  ballLayerMaxY
        }

        // 去掉隐式动画
        // 1.
        CATransaction.begin()
        CATransaction.setValue(kCFBooleanTrue, forKey: kCATransactionDisableActions)
        ballLayer.position = CGPointMake(ballLayer.position.x, ballLayerMinY)
        CATransaction.commit()
        // 2.
        // ballLayer.actions = ["position": nil]

        self.alpha = (fabs(currentOffsetY) - originalContentInset.top) / kRefreshHeight
        self.layer.hidden = false
    }

    func updateCircleLayerPosition() {
        circleLayer.position = CGPointMake(circleLayer.position.x, ballLayer.position.y - kSubviewEdage)
        circleLayer.hidden = false
        self.ballAnimation()
    }

    func ballAnimation() {
        let enlarge: CABasicAnimation = CABasicAnimation(keyPath: "transform.scale")
        enlarge.duration = 0.5
        enlarge.repeatCount = 1
        enlarge.removedOnCompletion = false
        enlarge.fromValue = 0.5
        enlarge.toValue   = 1

        let decrease: CABasicAnimation = CABasicAnimation(keyPath: "transform.scale")
        decrease.duration = 0.5
        decrease.repeatCount = 1
        decrease.removedOnCompletion = false
        decrease.fromValue = 1
        decrease.toValue   = 0.5
        decrease.beginTime = 0.5

        let position: CABasicAnimation = CABasicAnimation(keyPath: "position")
        position.duration = 1
        position.repeatCount = MAXFLOAT
        position.fromValue = NSValue(CGPoint: circleLayer.position)
        position.toValue = NSValue(CGPoint: CGPointMake(circleLayer.position.x, circleLayer.position.y + kTimeViewHeight))

        let group: CAAnimationGroup = CAAnimationGroup()
        group.duration = 1
        group.repeatCount = MAXFLOAT
        group.removedOnCompletion = false
        group.autoreverses = true
        group.animations = [enlarge, decrease, position]
        circleLayer.addAnimation(group, forKey: "CCEaseRefresh")
    }

    func updateTime() {
        switch refreshState {
            case .Normal, .Visible:
                lastUpdateTimeString = CCEaseDefaultTitle
            case .Targger:
                lastUpdateTimeString = CCEaseTriggertTitle
            case .Loading:
                lastUpdateTimeString = CCEaseLoadingTitle
        }
        self.setNeedsDisplay()
    }

    func calcTextXY() {
        if textX == 0.0 || textY == 0.0 {
            let options: NSStringDrawingOptions = .UsesLineFragmentOrigin
            let textSize = lastUpdateTimeString.boundingRectWithSize(CGSizeMake(CGFloat(MAXFLOAT), CGFloat(MAXFLOAT)), options: options, attributes: kTextAttribute, context: nil).size

            textY = self.frame.size.height - textSize.height - kSubviewEdage
            textX = (self.frame.size.width - textSize.width) / 2
        }
    }

    override func drawRect(rect: CGRect) {
        self.calcTextXY()
        lastUpdateTimeString.drawAtPoint(CGPointMake(textX, textY), withAttributes: kTextAttribute)
    }

    required init(coder aDecoder: NSCoder) {
        fatalError("init(coder:) has not been implemented")
    }

    override init(frame: CGRect) {
        super.init(frame: frame)
    }

    override func willMoveToSuperview(newSuperview: UIView?) {
        if (newSuperview == nil) {
            newSuperview?.removeObserver(self, forKeyPath: kContentOffset)
        }
    }
    
    func loadObserver() {
        targetView.addObserver(self, forKeyPath: kContentOffset, options: NSKeyValueObservingOptions.New, context: nil)
    }
    
    override func observeValueForKeyPath(keyPath: String?, ofObject object: AnyObject?, change: [String : AnyObject]?, context: UnsafeMutablePointer<Void>) {
        if keyPath == kContentOffset {
            if let result = change {

                currentOffsetY = (result[NSKeyValueChangeNewKey]?.CGPointValue)!.y

                if _refreshState == .Loading {return}
                let newOffsetThreshold = -originalContentInset.top
                if !targetView.dragging && _refreshState == .Targger {
                    refreshState = .Loading
                } else if currentOffsetY <= (2 * newOffsetThreshold) && targetView.dragging && _refreshState != .Targger {
                    refreshState = .Targger
                } else if (currentOffsetY < newOffsetThreshold && currentOffsetY > (2 * newOffsetThreshold)) {
                    refreshState = .Visible
                } else if (currentOffsetY >= newOffsetThreshold && _refreshState != .Normal) {
                    refreshState = .Normal
                }
            }
        }
    }
    
    deinit {
        print("CCEaseRefresh(Swift) deinit...");
        targetView?.removeObserver(self, forKeyPath: kContentOffset)
        targetView = nil
    }
    
    func loadComponent() {
        self.updateTime()
        self.calcTextXY()

        let selfViewWidth = self.frame.size.width;
        
        ballWidth = kBallImage.size.width   - kSubviewEdage / 2
        ballHeight = kBallImage.size.height - kSubviewEdage / 2
        ballLayer = CALayer()
        ballLayer.frame = CGRectMake((selfViewWidth - ballWidth) / 2, -ballHeight, ballWidth, ballHeight)
        ballLayer.position = CGPointMake(self.center.x, ballLayer.position.y)
        ballLayer.contents = kBallImage.CGImage
        self.layer.addSublayer(ballLayer)
        
        circleWidth = kCircleImage.size.width
        cricleHeight = kCircleImage.size.height
        circleLayer = CALayer()
        circleLayer.frame = CGRectMake((selfViewWidth - circleWidth) / 2, 0, circleWidth, cricleHeight)
        circleLayer.contents = kCircleImage.CGImage
        circleLayer.hidden = true
        
        defaultBallY = ballLayer.frame.origin.y
        
        self.alpha = 0
        self.layer.addSublayer(ballLayer)
        self.layer.insertSublayer(circleLayer, below: ballLayer)
    }

    func beginRefreshing() {
        if _refreshState != .Loading {
            refreshState = .Loading;
        }
    }

    func endRefreshing() {
        if _refreshState != .Normal {
            refreshState = .Normal
            self.alpha = 0.0
            self.hidden = true
            circleLayer.removeAnimationForKey("CCEaseRefresh")
            circleLayer.hidden = true
            ballLayer.position = CGPointMake(ballLayer.position.x, defaultBallY)
        }
    }
}

enum CCEaseError: ErrorType {
    case OJCNotExist
}

extension CCEaseRefresh {

    // Swift错误处理: http://www.cocoachina.com/swift/20150623/12231.html
    convenience init(scrollView: UIScrollView) /*throws*/ {
        /*guard scrollView == true  else {
            throw CCEaseError.OJCNotExist
        }*/
        self.init()

        targetView = scrollView
        originalContentInset = scrollView.contentInset
        self.frame = CGRectMake(0, originalContentInset.top, targetView.frame.size.width, kRefreshHeight)
        self.backgroundColor = UIColor.clearColor()
        targetView.backgroundColor = UIColor.clearColor()
        targetView.superview?.insertSubview(self, belowSubview: targetView)
        
        self.loadObserver()
        self.loadComponent()
    }
}




